### PR TITLE
# 198 Support Passing Command Line Arguments To Driver

### PR DESCRIPTION
**Pull request by oomelianchuk**
Added driver arguments support for Chrome, Firefox, IE and Safari (Opera and PhantomJs won't be supported by Selenium in the next version, so the is no point to develop a feature for those browsers). As the most popular and, what's more important, cross-platform browsers are Chrome and Firefox, the solution was tested for these browsers, both manually and via unit tests. 

PS: It would be good to know, that the solution also works for ie and safari, so feel free to try the feature out on your platform, especially if you have access to safari and ie browsers. Thanks!

### CODE REVIEW

File: src/test/java/com/xceptance/neodymium/tests/DriverArgumentsTest.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+@Browser("FF_headless")
+public class DriverArgumentsTest extends NeodymiumTest
+{
+    private static String randomLogFileName = "target/" + UUID.randomUUID().toString() + ".log";
+
+    @BeforeClass
+    public static void createSettings()
+    {
+        Map<String, String> properties1 = new HashMap<>();
+        properties1.put("neodymium.webDriver.chrome.driverArguments", "--silent ; --log-path=" + randomLogFileName);
```
Comment (wurzelkuchen) [inline src/test/java/com/xceptance/neodymium/tests/DriverArgumentsTest.java:29]: The Issue has a list with possible arguments, but the issue is quite old, please check if everything is still valid for current browsers.

File: src/test/java/com/xceptance/neodymium/tests/DriverArgumentsTest.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import org.junit.runner.RunWith;
+
+import com.codeborne.selenide.Selenide;
+import com.xceptance.neodymium.NeodymiumRunner;
+import com.xceptance.neodymium.module.statement.browser.multibrowser.Browser;
+
+@RunWith(NeodymiumRunner.class)
+@Browser("Chrome_headless")
+@Browser("FF_headless")
+public class DriverArgumentsTest extends NeodymiumTest
```
Comment (wurzelkuchen) [inline src/test/java/com/xceptance/neodymium/tests/DriverArgumentsTest.java:21]: please add a test for an empty config string as well.

**Review by wurzelkuchen — CHANGES_REQUESTED** [review]

File: src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/wrappers/ChromeBuilder.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        firefoxBinary = findDefaultExecutable();
+        port = PortProber.findFreePort();
+        this.arguments = arguments;
+        try
+        {
+            return new ChromeDriverService(firefoxBinary, port, createArgs(), ImmutableMap.copyOf(System.getenv()));
+        }
+        catch (IOException e)
+        {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/wrappers/ChromeBuilder.java]: Removed during refactoring

File: src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/wrappers/GeckoBuilder.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                }
+                else
+                {
+                    try
+                    {
+                        service.sendOutputTo(new FileOutputStream(firefoxLogFile));
+                    }
+                    catch (IOException e)
+                    {
+                        e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/wrappers/GeckoBuilder.java]: Removed during refactoring

[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/wrappers/GeckoBuilder.java]: Removed during refactoring

File: src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/wrappers/GeckoBuilder.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        var logPathArgs = arguments.stream().filter(arg -> arg.contains("--log-path")).collect(Collectors.toList());
+        this.arguments.removeAll(logPathArgs);
+        var service = createDriverService(firefoxBinary, port, createArgs(), ImmutableMap.copyOf(System.getenv()));
+
+        if (!logPathArgs.isEmpty())
+        {
+            String firefoxLogFile = logPathArgs.get(0).replaceAll("--log-path=", "").trim();
+            if (firefoxLogFile != null)
+            {
+                if ("/dev/stdout".equals(firefoxLogFile))
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/wrappers/GeckoBuilder.java]: after refactoring delegated to Selenium (the "/dev/stdout" and "/dev/stderr", etc. are cross-platform in Selenium, these are only indicators for where to send logs output, in the end it's done via System.out or System.err)

File: src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/wrappers/GeckoBuilder.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        return service;
+    }
+
+    @Override
+    protected ImmutableList<String> createArgs()
+    {
+        ImmutableList.Builder<String> argsBuilder = ImmutableList.builder();
+        // argsBuilder.addAll(super.createArgs());
+        argsBuilder.add(String.format("--port=%d", port));
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/wrappers/GeckoBuilder.java]: adjusted

File: src/main/java/com/xceptance/neodymium/util/NeodymiumConfiguration.java [hunk 7]
```diff
@@ -242,15 +242,28 @@ public interface NeodymiumConfiguration extends Mutable
     @Key("neodymium.webDriver.chrome.pathToDriverServer")
     public String getChromeDriverPath();
 
-    @Key("neodymium.webDriver.edge.pathToDriverServer")
-    public String getEdgeDriverPath();
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/util/NeodymiumConfiguration.java:246]: thanks for pointing this out! It should not be removed, furthermore, it missed implementation as IE and Edge now require different setups. Setup for Edge implemented. The setup for IE stays the same but it's now in investigation if we still need to support IE

File: src/test/java/com/xceptance/neodymium/tests/DriverArgumentsTest.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+{
+    private static String randomLogFileName = "target/" + UUID.randomUUID().toString() + ".log";
+
+    @BeforeClass
+    public static void createSettings()
+    {
+        Map<String, String> properties1 = new HashMap<>();
+        properties1.put("neodymium.webDriver.chrome.driverArguments", "--silent ; --log-path=" + randomLogFileName);
+        properties1.put("neodymium.webDriver.firefox.driverArguments", "--log ; fatal ; --log-path=" + randomLogFileName);
+        File tempConfigFile1 = new File("./config/dev-neodymium.properties");
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/tests/DriverArgumentsTest.java]: done

File: src/test/java/com/xceptance/neodymium/tests/DriverArgumentsTest.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import com.codeborne.selenide.Selenide;
+import com.xceptance.neodymium.NeodymiumRunner;
+import com.xceptance.neodymium.module.statement.browser.multibrowser.Browser;
+
+@RunWith(NeodymiumRunner.class)
+@Browser("Chrome_headless")
+@Browser("FF_headless")
+public class DriverArgumentsTest extends NeodymiumTest
+{
+    private static String randomLogFileName = "target/" + UUID.randomUUID().toString() + ".log";
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/tests/DriverArgumentsTest.java]: done

File: src/test/java/com/xceptance/neodymium/tests/DriverArgumentsTest.java (hunk 2 above)
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/tests/DriverArgumentsTest.java]: implicity covered with other tests, adding tests for invalid values will test browser behavior but not our code

**Review by wurzelkuchen — APPROVED** [review]
